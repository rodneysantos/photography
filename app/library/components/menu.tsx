import { useState } from "react";

interface MenuItem {
	// The id of the item.
	id: number;

	// The name of the item.
	name: Item;
}

const activeMenuClasses = "bg-neutral-900 text-neutral-100";
const menuItems: MenuItem[] = [
	{ id: 0, name: "black and white" },
	{ id: 1, name: "color" },
	{ id: 2, name: "low-key" },
];

export type Item = "black and white" | "color" | "low-key";

export interface MenuProps {
	// The class name to apply to the div element.
	className?: string;

	// The ids of the active items.
	defaultActiveItems: Item[];

	// The callback to call when the active items change.
	onActiveItemsChange: (activeItems: Item[]) => void;
}

export const Menu = ({
	defaultActiveItems,
	onActiveItemsChange,
	className,
}: MenuProps) => {
	const [activeItems, setActiveItems] = useState(defaultActiveItems);

	// Toggle the active state of an item.
	// If the item is active, remove it from the active items.
	// If the item is not active, add it to the active items.
	const handleToggleItem = (name: Item) => {
		const updatedActiveItems = activeItems.includes(name)
			? activeItems.filter((item) => item !== name)
			: [...activeItems, name];

		setActiveItems(updatedActiveItems);
		onActiveItemsChange(updatedActiveItems);
	};

	return (
		<div data-testid="menu" className={className}>
			<div className="flex flex-row justify-center">
				{menuItems.map((item) => {
					const isActive = activeItems.includes(item.name);
					const className = isActive ? activeMenuClasses : "";

					return (
						<div
							key={item.id}
							data-testid={item.name}
							className={`px-2 py-1 mx-2 text-center border-2 cursor-pointer border-neutral-900 ${className}`}
							onClick={() => handleToggleItem(item.name)}
						>
							{item.name}
						</div>
					);
				})}
			</div>
		</div>
	);
};
