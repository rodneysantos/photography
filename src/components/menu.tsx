import { createSignal } from "solid-js";
import { JSX } from "solid-js/jsx-runtime";

interface Menu {
	// The id of the item.
	id: number;

	// The key of the item.
	key: MenuKey;

	// The name of the item.
	name: "Black and White" | "Color" | "Low-Key";
}

export type MenuKey = "bnw" | "color" | "low-key";

const menuItems: Menu[] = [
	{ id: 0, key: "bnw", name: "Black and White" },
	{ id: 1, key: "color", name: "Color" },
	{ id: 2, key: "low-key", name: "Low-Key" },
];

export interface MenuProps extends JSX.HTMLAttributes<HTMLDivElement> {
	// The ids of the active items.
	defaultActiveItems: MenuKey[];

	// The callback to call when the active items change.
	onActiveItemsChange: (activeItems: MenuKey[]) => void;

	// Whether or not to show the menu.
	show: boolean;
}

export const Menu = (props: MenuProps) => {
	const [activeItems, setActiveItems] = createSignal(props.defaultActiveItems);

	// Toggle the active state of an item.
	// If the item is active, remove it from the active items.
	// If the item is not active, add it to the active items.
	function handleToggleItem(key: MenuKey) {
		const updatedActiveItems = activeItems().includes(key)
			? activeItems().filter((item) => item !== key)
			: [...activeItems(), key];

		setActiveItems(updatedActiveItems);
		props.onActiveItemsChange(updatedActiveItems);
	}

	return (
		<div
			data-testid="menu"
			class="transition-opacity will-change-auto"
			classList={{ "opacity-0": !props.show }}
		>
			<div class="flex flex-row justify-center">
				{menuItems.map((item) => {
					const isActive = activeItems().includes(item.key);

					return (
						<div
							data-testid={item.key}
							onClick={() => handleToggleItem(item.key)}
							class="px-2 py-1 mx-2 text-center border-2 cursor-pointer border-neutral-900"
							classList={{ "bg-neutral-900 text-neutral-100": isActive }}
						>
							{item.name}
						</div>
					);
				})}
			</div>
		</div>
	);
};
