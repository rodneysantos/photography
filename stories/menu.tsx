import { Menu } from "../app/library/components/menu";

export const MenuSB = () => {
	return <Menu />;
};
