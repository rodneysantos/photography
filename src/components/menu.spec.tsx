import { cleanup, render } from "@solidjs/testing-library";
import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import { Menu } from "./menu";

describe("menu", () => {
	const onActiveItemsChange = vi.fn();

	beforeEach(() => {
		onActiveItemsChange.mockReset();
	});

	afterEach(cleanup);

	test("renders correctly", () => {
		// arrange
		const { asFragment } = render(() => (
			<Menu
				show={true}
				defaultActiveItems={[]}
				onActiveItemsChange={onActiveItemsChange}
			/>
		));

		// assert
		expect(asFragment()).toMatchSnapshot();
	});

	test("highlights the items on render", () => {
		// arrange
		const { getByTestId } = render(() => (
			<Menu
				show={true}
				defaultActiveItems={["bnw", "low-key"]}
				onActiveItemsChange={onActiveItemsChange}
			/>
		));

		// assert
		expect(getByTestId("bnw").className).toContain("bg-neutral-900");
		expect(getByTestId("low-key").className).toContain("bg-neutral-900");
	});

	test("higlights an item on click", () => {
		// arrange
		const { getByTestId } = render(() => (
			<Menu
				show={true}
				defaultActiveItems={[]}
				onActiveItemsChange={onActiveItemsChange}
			/>
		));

		// act
		getByTestId("color").click();

		// assert
		expect(getByTestId("color").className).toContain("bg-neutral-900");
	});

	test("unhiglights an item on click", () => {
		// arrange
		const { getByTestId } = render(() => (
			<Menu
				show={true}
				defaultActiveItems={["bnw"]}
				onActiveItemsChange={onActiveItemsChange}
			/>
		));

		// act
		getByTestId("bnw").click();

		// assert
		expect(getByTestId("bnw").className).not.toContain("bg-neutral-900");
	});

	test("emits the active items on click", () => {
		// arrange
		const { getByTestId } = render(() => (
			<Menu
				show={true}
				defaultActiveItems={[]}
				onActiveItemsChange={onActiveItemsChange}
			/>
		));

		// act
		getByTestId("color").click();
		getByTestId("low-key").click();

		// assert
		expect(onActiveItemsChange).toHaveBeenCalledTimes(2);
		expect(onActiveItemsChange).toHaveBeenCalledWith(["color"]);
		expect(onActiveItemsChange).toHaveBeenCalledWith(["color", "low-key"]);
	});
});
