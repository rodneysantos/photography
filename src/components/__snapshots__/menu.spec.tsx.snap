// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`menu > renders correctly 1`] = `"<div data-testid=\\"menu\\" class=\\"transition-opacity will-change-auto\\"><div class=\\"flex flex-row justify-center\\"><div class=\\"px-2 py-1 mx-2 text-center border-2 cursor-pointer border-neutral-900\\" data-testid=\\"bnw\\">Black and White</div><div class=\\"px-2 py-1 mx-2 text-center border-2 cursor-pointer border-neutral-900\\" data-testid=\\"color\\">Color</div><div class=\\"px-2 py-1 mx-2 text-center border-2 cursor-pointer border-neutral-900\\" data-testid=\\"low-key\\">Low-Key</div></div></div>"`;
